walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,53 +15,122 @@
-        popedRange = busyIntervals.pop();
-        busyIntervals.push(mergeTimeRanges(popedRange, timeRange));
+    if(importantRanges.size() > 0){
+      busyRanges.push(importantRanges.get(0));
+      for(TimeRange timeRange : importantRanges){
+        if(timeRange.overlaps(busyRanges.peek())){
+          popedRange = busyRanges.pop();
+          busyRanges.push(mergeTimeRanges(popedRange, timeRange));
+        }else{
+          busyRanges.push(timeRange);
+        }
+      }
+    }
+  
+    optionalBusyRanges.push(optionalImportantRanges.get(0));
+    for(TimeRange timeRange : optionalImportantRanges){
+      if(timeRange.overlaps(optionalBusyRanges.peek())){
+        popedRange = optionalBusyRanges.pop();
+        optionalBusyRanges.push(mergeTimeRanges(popedRange, timeRange));
       }else{
-        busyIntervals.push(timeRange);
+        optionalBusyRanges.push(timeRange);
       }
     }
 
+    //looking for free spaces among merged intervals
     int duration;
     //dummy TimeRanges to provide limits to the entire space
-    busyIntervals.add(0, new TimeRange(TimeRange.START_OF_DAY, TimeRange.START_OF_DAY));
-    busyIntervals.add(new TimeRange(TimeRange.END_OF_DAY+1, TimeRange.END_OF_DAY+1));
+    busyRanges.add(0, new TimeRange(TimeRange.START_OF_DAY, TimeRange.START_OF_DAY));
+    busyRanges.add(new TimeRange(TimeRange.END_OF_DAY+1, TimeRange.END_OF_DAY+1));
+    optionalBusyRanges.add(0, new TimeRange(TimeRange.START_OF_DAY, TimeRange.START_OF_DAY));
+    optionalBusyRanges.add(new TimeRange(TimeRange.END_OF_DAY+1, TimeRange.END_OF_DAY+1));
     //this loop calculates size of free space between two subsequent intervals, and checks if size meets requirements
     //if so it adds this space to freeIntervals that stores possible place for requested event
-    for(int i = 0; i < busyIntervals.size()-1; i++){
-      duration = busyIntervals.get(i+1).start()-busyIntervals.get(i).end();
+    for(int i = 0; i < busyRanges.size()-1; i++){
+      duration = busyRanges.get(i+1).start()-busyRanges.get(i).end();
+      if(duration > 0 && duration >= request.getDuration())
+        freeRanges.add(new TimeRange(busyRanges.get(i).end(), busyRanges.get(i+1).start()-busyRanges.get(i).end()));
+    }
+    for(int i = 0; i < optionalBusyRanges.size()-1; i++){
+      duration = optionalBusyRanges.get(i+1).start()-optionalBusyRanges.get(i).end();
       if(duration > 0 && duration >= request.getDuration())
-        freeIntervals.add(new TimeRange(busyIntervals.get(i).end(), busyIntervals.get(i+1).start()-busyIntervals.get(i).end()));
+        freeRangesWithOptionals.add(new TimeRange(optionalBusyRanges.get(i).end(), optionalBusyRanges.get(i+1).start()-optionalBusyRanges.get(i).end()));
     }
 
-    return freeIntervals;
+    //if with optional attendees there is no room for the meeting we return freeRangesWithoutOptionals
+    if(freeRangesWithOptionals.size() == 0) return freeRanges;
+    else return freeRangesWithOptionals;
```

Comment:
Here it's always calculating both `freeRanges` & `freeRangesWithOptionals`, even if when non-empty `freeRangesWithOptionals` is found, the calculation for `freeRanges` is not wasted.

Can we do better?

Reply:
Oh yes! We can first calculate free space with optional attendees and only if there is no space with such restriction calculate space without optional attendees.